### ISSUE TITLE
[EKS] [request]: Native, granular ECR permissions

### ISSUE THREAD

**Issue by valorl** [aws/containers-roadmap#2133]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Would be great to have a more granular permission mechansim to enforce ECR pull access on a Pod or Namespace level instead of EC2 level. 

For example, there could be an additional IRSA-like annotation on a ServiceAccount to specify the IAM Role that should be used for the image pull and then maybe some controller could continously impersonate that role renew an imagePullSecret for that SA.

**Which service(s) is this request for?**
EKS / ECR

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Currently, EKS -> ECR permissions are scoped to nodes. This means that as long as someone has access to create a Pod on a given node, they can run _any_ image the node's allowed to pull, even if that image is owned by another team.

This is generally not a problem with static credentials and external registries: As long as the host supports scoping users/keys, granular keys can be provisioned and used in image pull secrets. However, this is not an option for ECR, as static credentials are short-lived.

**Are you currently working around this issue?**
Potential workarounds are:
- Design node groups around ECR permission boundaries (e.g. team/namespace)
- Use K8S API admission (Kyverno, Gatekeeper, etc) on create Pod requests to reject/allow images more granularly

**Comment by mikestef9** [comment]
Hi @valorl 

Images are pulled by kubelet which runs on the worker node, not by pods themselves.

The closest thing to this upstream that never actually happened 2ys ago, doc details below is
https://docs.google.com/document/d/1qX7FoXBxWrPfUD55IDQMyeDQMH-HzH8vSx6-WsLNr8o/edit?resourcekey=0-iLUGguVKXPeHMNDwEzBMbw#heading=h.foqb0s2snh9q

If it had been implemented, essentially this would let you use a secondary ClientID/Audience on the same token so it could be leveraged for pull creds.

But this is a request that would need to be handled upstream, it's not related to IRSA.

**Comment by valorl** [comment]
Right, I understand image pulls are tied to kubelet, but currently it's still possible to scope them to Pod level using separate image pull credentials. 

As I mentioned, I'm thinking that it would technically be achievable with a controller that's responsible for provisioning and renewing ECR tokens into image pull secrets, based on configurable IAM roles per Pod/SA.

That being said, it may too "hacky" for this roadmap.
